src/hooks/useData.ts

```diff
@@ -0,0 +1,51 @@
+import { useEffect, useCallback } from 'react';
+import { Product } from '../components/ProductCardList/product.type';
+import { CartItem } from '../components/ShoppingCartModal/cart.type';
+import { useDataContext } from '../contexts/useDataContext';
+
+interface UseDataOptions<T> {
+  key: string;
+  endpoint: string;
+  fetchFunction: (endpoint: string) => Promise<T>;
+}
+
+export const useData = <T extends CartItem[] | Product[] | null>({
+  key,
+  endpoint,
+  fetchFunction,
+}: UseDataOptions<T>) => {
+  const { state, setData, setLoading, setError } = useDataContext();
+
+  const fetchData = useCallback(async () => {
+    setLoading(key, true);
+    setError(key, '');
+
+    try {
+      const data = await fetchFunction(endpoint);
+      setData(key, data ?? []);
+    } catch (error) {
+      const errorMessage =
+        error instanceof Error ? error.message : 'Unknown error';
+      setError(key, errorMessage);
+    }
+  }, [key, endpoint, setData, setLoading, setError, fetchFunction]);
```

Comment:
이런 모습으로 관리하는걸 지양하기 위해 리듀서를 사용했던건데.. 지금은 리듀서를 쓰는 의미가 조금 퇴색되는 것 같아요 ㅠㅠ

이렇게 쓰는 모습을 상상해보시면 좋겠습니다.


```suggestion
  const { state, load, start, finish, onError } = useDataContext(key);

  const fetchData = useCallback(async () => {
    start();
    fetchFunction(endpoint)
      .then(load)
      .catch(onError)
      .finally(finish)
    }
  }, [endpoint, start, load, onError, finish]);
```

지금은 set으로 하나하나 바꿔주고 있는데,
reducer 로 api 요청 흐름을 추상화해서 관리하는거죠 ㅎㅎ